synapse/util/caches/response_cache.py

```diff
@@ -112,6 +117,31 @@ def remove(r):
         result.addBoth(remove)
         return result.observe()
 
+    def add_conditional(self, key: T, conditional: Callable[[Any], bool]):
+        self.pending_conditionals.setdefault(key, set()).add(conditional)
+
+    def wrap_conditional(
+        self,
+        key: T,
+        should_cache: Callable[[Any], bool],
+        callback: "Callable[..., Any]",
+        *args: Any,
+        **kwargs: Any
+    ) -> defer.Deferred:
+        """The same as wrap(), but adds a conditional to the final execution.
+
+        When the final execution completes, *all* conditionals need to return True for it to properly cache,
+        else it'll not be cached in a timed fashion.
+        """
+
+        # See if there's already a result on this key that hasn't yet completed. Due to the single-threaded nature of
+        # python, adding a key immediately in the same execution thread will not cause a race condition.
+        result = self.get(key)
+        if not result or isinstance(result, defer.Deferred) and not result.called:
+            self.add_conditional(key, should_cache)
```

Comment:
is it correct that the conditional is added when `not result.called` ?  I'd argue that cachability of the result should be determined by the `callback` being used, rather than any subsequent `callbacks` (which are discarded).

Reply:
operator precedence is `or` -> `and` -> `not`, so this is actually
`((not result) or isinstance(result, defer.Deferred)) and (not result.called)`

I had to check after your previous comment, it could actually be `if not result and not result.called`, but because `result` can be a plain return value, i need to insert the check to `isinstance(..., defer.Deferred)`

Yes, the conditional is added to be executed after the result is called, i do agree that now i realise that any future calls can be like "hey, thjis shouldn't be cached anyways because the returned value doesn't agree with what i already got from somewhere else", in which case I need to rewrite this to execute the conditional locally whenever it's called, and evict the cache when it's not valid according to the new conditional. (the `callLater` already taking place would simply execute anyways, no way to properly garbage collect that at that point to my knowledge, if it was an asyncio `Task`, i could also map it somewhere to then `.cancel()` it, if this is possible with `callLater` handles or something, please tell me)

Reply:
as discussed at considerable length in the room: I don't agree with your logic here. I think that supporting multiple `should_cache` callbacks per key significantly complicates the implentation, and semantically is dubious at best.

Reply:
So i can do three things here (sorry to repeat again);
1. Only the *very first* call will add the conditional (replacing the `dict[T, set[Callable]]` with just a plain `dict[T, Callable]` mapping).
2. The first callers (to call `wrap` before the call completes) will be able to contribute a `conditional`, this is how it's currently implemented.
3. Any subsequent call to `wrap_conditional`, even after the result has been set, can veto and evict the cache.

I didn't get a clear signal if it should be `1` or `2`, that's why i'm asking again.

Reply:
Alright, i'll just make it a simple `Dict[T, Callable]` mapping then

Reply:
Actually, more discussion in #9522 on this.